dolphin/drivers/manager.py

```diff
@@ -26,57 +27,76 @@
+        :param invoke_on_load: Boolean to decide whether to return the
+            driver object.
+        :type invoke_on_load: bool
+        :param cache_on_load: Boolean to decide whether save driver object
+            in driver_factory when generating a new driver object.
+            It takes effect when invoke_on_load is True.
+        :type invoke_on_load: bool
+        :param kwargs: Parameters from access_info.
+        """
+        if not invoke_on_load:
+            return self._get_driver_cls(**kwargs)
+        else:
+            return self._get_driver_obj(context, cache_on_load, **kwargs)
 
-    def create_driver(self, context, **kwargs):
-        storage_id = kwargs.get('storage_id', None)
-        driver = self._init_driver(context, **kwargs)
+    def update_driver(self, storage_id, driver):
         self.driver_factory[storage_id] = driver
 
-        return driver
+    def remove_driver(self, storage_id):
+        """Clear driver instance from driver factory."""
+        self.driver_factory.pop(storage_id, None)
+
+    def _get_driver_obj(self, context, cache_on_load=True, **kwargs):
+        if not cache_on_load or not kwargs.get('storage_id'):
+            cls = self._get_driver_cls(**kwargs)
+            return cls(**kwargs)
 
-    def get_driver(self, context, storage_id):
-        driver = self.driver_factory.get(storage_id, None)
-        if not driver:
-            with self._instance_lock:
-                driver = self.driver_factory.get(storage_id, None)
-                if not driver:
-                    access_info = helper.get_access_info(context, storage_id)
-                    driver = self._init_driver(context, **access_info)
-                    self.driver_factory[storage_id] = driver
+        if kwargs['storage_id'] in self.driver_factory:
+            return self.driver_factory[kwargs['storage_id']]
 
-        return driver
+        with self._instance_lock:
+            if kwargs['storage_id'] in self.driver_factory:
+                return self.driver_factory[kwargs['storage_id']]
 
-    def remove_driver(self, context, storage_id):
-        """Clear driver instance from driver factory."""
-        self.driver_factory.pop(storage_id, None)
+            access_info = copy.deepcopy(kwargs)
+            storage_id = access_info.pop('storage_id')
+            if access_info:
+                cls = self._get_driver_cls(**kwargs)
+                driver = cls(**kwargs)
+            else:
+                access_info = helper.get_access_info(context, storage_id)
+                cls = self._get_driver_cls(**access_info)
+                driver = cls(**access_info)
+
+            self.driver_factory[storage_id] = driver
+            return driver
+
+    def _get_driver_cls(self, **kwargs):
+        """Get driver class from entry points."""
+        name = '%s %s' % (kwargs.get('vendor'), kwargs.get('model'))
+        if name in self.names():
+            return self[name].plugin
+        else:
```

Comment:
optimization suggestion:
else statment can be removed.

Reply:
done